File: backend/app/models/clinical/opd.py
"""clinical schema OPD tables — visits, queue, vitals, notes, prescriptions, referrals."""

from __future__ import annotations

import uuid
from datetime import date, datetime
from decimal import Decimal

from sqlalchemy import Boolean, Date, DateTime, Integer, Numeric, SmallInteger, String, Text
from sqlalchemy.dialects.postgresql import UUID as PG_UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import TenantAuditableEntity


class OpdVisit(TenantAuditableEntity):
    __tablename__ = "opd_visits"
    __table_args__ = {"schema": "clinical"}

    visit_number: Mapped[str] = mapped_column(String(20), nullable=False)
    patient_id: Mapped[uuid.UUID] = mapped_column(PG_UUID(as_uuid=True), nullable=False)
    doctor_id: Mapped[uuid.UUID] = mapped_column(PG_UUID(as_uuid=True), nullable=False)
    appointment_id: Mapped[uuid.UUID | None] = mapped_column(PG_UUID(as_uuid=True), nullable=True)
    location_id: Mapped[uuid.UUID | None] = mapped_column(PG_UUID(as_uuid=True), nullable=True)
    visit_date: Mapped[date] = mapped_column(Date, nullable=False)
    visit_type: Mapped[str] = mapped_column(String(20), nullable=False)
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="waiting")
    token_number: Mapped[int | None] = mapped_column(Integer, nullable=True)
    chief_complaint: Mapped[str | None] = mapped_column(Text, nullable=True)
    started_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    completed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)


class OpdQueue(TenantAuditableEntity):
    __tablename__ = "opd_queue"
    __table_args__ = {"schema": "clinical"}

    opd_visit_id: Mapped[uuid.UUID] = mapped_column(PG_UUID(as_uuid=True), nullable=False)
    doctor_id: Mapped[uuid.UUID] = mapped_column(PG_UUID(as_uuid=True), nullable=False)
    token_number: Mapped[int] = mapped_column(Integer, nullable=False)
    queue_date: Mapped[date] = mapped_column(Date, nullable=False)
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="waiting")
    priority: Mapped[str] = mapped_column(String(10), nullable=False, default="normal")
    called_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)


class OpdVitals(TenantAuditableEntity):
    __tablename__ = "opd_vitals"
    __table_args__ = {"schema": "clinical"}

    opd_visit_id: Mapped[uuid.UUID] = mapped_column(PG_UUID(as_uuid=True), nullable=False)
    recorded_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    blood_pressure_systolic: Mapped[int | None] = mapped_column(SmallInteger, nullable=True)
    blood_pressure_diastolic: Mapped[int | None] = mapped_column(SmallInteger, nullable=True)
    pulse_rate: Mapped[int | None] = mapped_column(SmallInteger, nullable=True)
    temperature: Mapped[Decimal | None] = mapped_column(Numeric(4, 1), nullable=True)
    respiratory_rate: Mapped[int | None] = mapped_column(SmallInteger, nullable=True)
    spo2: Mapped[int | None] = mapped_column(SmallInteger, nullable=True)
    weight_kg: Mapped[Decimal | None] = mapped_column(Numeric(5, 2), nullable=True)
    height_cm: Mapped[Decimal | None] = mapped_column(Numeric(5, 1), nullable=True)
    bmi: Mapped[Decimal | None] = mapped_column(Numeric(4, 1), nullable=True)
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)


class OpdClinicalNote(TenantAuditableEntity):
    __tablename__ = "opd_clinical_notes"
    __table_args__ = {"schema": "clinical"}

    opd_visit_id: Mapped[uuid.UUID] = mapped_column(PG_UUID(as_uuid=True), nullable=False)
    note_type: Mapped[str] = mapped_column(String(30), nullable=False)
    content: Mapped[str] = mapped_column(Text, nullable=False)
    icd_code: Mapped[str | None] = mapped_column(String(10), nullable=True)
    icd_description: Mapped[str | None] = mapped_column(String(255), nullable=True)
    is_final: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)


class OpdPrescription(TenantAuditableEntity):
    __tablename__ = "opd_prescriptions"
    __table_args__ = {"schema": "clinical"}

    opd_visit_id: Mapped[uuid.UUID] = mapped_column(PG_UUID(as_uuid=True), nullable=False)
    patient_id: Mapped[uuid.UUID] = mapped_column(PG_UUID(as_uuid=True), nullable=False)
    doctor_id: Mapped[uuid.UUID] = mapped_column(PG_UUID(as_uuid=True), nullable=False)
    prescription_number: Mapped[str] = mapped_column(String(20), nullable=False)
    prescribed_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="active")
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)


class OpdPrescriptionItem(TenantAuditableEntity):
    __tablename__ = "opd_prescription_items"
    __table_args__ = {"schema": "clinical"}

    prescription_id: Mapped[uuid.UUID] = mapped_column(PG_UUID(as_uuid=True), nullable=False)
    medicine_id: Mapped[uuid.UUID | None] = mapped_column(PG_UUID(as_uuid=True), nullable=True)
    medicine_name: Mapped[str] = mapped_column(String(255), nullable=False)
    dosage: Mapped[str] = mapped_column(String(100), nullable=False)
    frequency: Mapped[str] = mapped_column(String(100), nullable=False)
    duration: Mapped[str] = mapped_column(String(100), nullable=False)
    route: Mapped[str | None] = mapped_column(String(50), nullable=True)
    instructions: Mapped[str | None] = mapped_column(Text, nullable=True)
    quantity: Mapped[int | None] = mapped_column(Integer, nullable=True)


class OpdReferral(TenantAuditableEntity):
    __tablename__ = "opd_referrals"
    __table_args__ = {"schema": "clinical"}

    opd_visit_id: Mapped[uuid.UUID] = mapped_column(PG_UUID(as_uuid=True), nullable=False)
    patient_id: Mapped[uuid.UUID] = mapped_column(PG_UUID(as_uuid=True), nullable=False)
    referring_doctor_id: Mapped[uuid.UUID] = mapped_column(PG_UUID(as_uuid=True), nullable=False)
    referred_doctor_id: Mapped[uuid.UUID | None] = mapped_column(PG_UUID(as_uuid=True), nullable=True)
    referral_type: Mapped[str] = mapped_column(String(20), nullable=False)
    external_facility: Mapped[str | None] = mapped_column(String(255), nullable=True)
    reason: Mapped[str] = mapped_column(Text, nullable=False)
    urgency: Mapped[str] = mapped_column(String(10), nullable=False, default="routine")
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="pending")

File: frontend/src/api/types/opd.ts
export type OpdQueueStatus =
  | "waiting"
  | "called"
  | "in_consultation"
  | "completed"
  | "skipped";

export type OpdQueuePriority = "normal" | "urgent" | "emergency";

export type OpdVisitStatus =
  | "waiting"
  | "in_consultation"
  | "completed"
  | "cancelled"
  | "no_show";

export interface OpdQueueEntry {
  id: string;
  opd_visit_id: string;
  doctor_id: string;
  patient_id: string;
  patient_name: string | null;
  token_number: number;
  queue_date: string;
  status: OpdQueueStatus;
  priority: OpdQueuePriority;
  chief_complaint: string | null;
  called_at: string | null;
  visit_status: OpdVisitStatus | null;
}

export interface OpdQueueBoard {
  doctor_id: string;
  queue_date: string;
  current_token: number | null;
  waiting_count: number;
  entries: OpdQueueEntry[];
}

export interface OpdQueuePoll extends OpdQueueBoard {
  etag: string;
  poll_interval_seconds: number;
  changed: boolean;
}

export interface OpdQueueSkipPayload {
  reason: string;
}

export interface OpdQueueUpdatePayload {
  priority?: OpdQueuePriority;
  position?: number;
}

export type OpdVisitType = "walk_in" | "appointment";

export type OpdNoteType = "examination" | "diagnosis" | "plan" | "general";

export type OpdPrescriptionStatus =
  | "active"
  | "dispensed"
  | "partially_dispensed"
  | "cancelled";

export type MedicineRoute = "oral" | "topical" | "iv" | "im" | "sc" | "inhalation" | "other";

export interface OpdVisit {
  id: string;
  tenant_id: string;
  visit_number: string;
  patient_id: string;
  patient_name: string | null;
  patient_mrn: string | null;
  doctor_id: string;
  doctor_name: string | null;
  appointment_id: string | null;
  location_id: string | null;
  location_name: string | null;
  visit_date: string;
  visit_type: OpdVisitType;
  status: OpdVisitStatus;
  token_number: number | null;
  chief_complaint: string | null;
  started_at: string | null;
  completed_at: string | null;
  queue_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface OpdVisitDetail extends OpdVisit {
  vitals_count: number;
  notes_count: number;
  prescriptions_count: number;
}

export interface OpdVisitCompleteResult {
  id: string;
  status: OpdVisitStatus;
  completed_at: string;
  billing_invoice_id: string | null;
}

export interface OpdVisitCompletePayload {
  finalize_notes?: boolean;
  create_billing_draft?: boolean;
  follow_up_notes?: string | null;
}

export interface OpdVitals {
  id: string;
  opd_visit_id: string;
  recorded_at: string;
  recorded_by_user_id: string | null;
  blood_pressure_systolic: number | null;
  blood_pressure_diastolic: number | null;
  pulse_rate: number | null;
  temperature: number | null;
  temperature_unit: string | null;
  respiratory_rate: number | null;
  spo2: number | null;
  weight_kg: number | null;
  height_cm: number | null;
  bmi: number | null;
  notes: string | null;
}

export interface OpdVitalsCreatePayload {
  blood_pressure_systolic?: number | null;
  blood_pressure_diastolic?: number | null;
  pulse_rate?: number | null;
  temperature?: number | null;
  respiratory_rate?: number | null;
  spo2?: number | null;
  weight_kg?: number | null;
  height_cm?: number | null;
  notes?: string | null;
}

export interface OpdClinicalNote {
  id: string;
  opd_visit_id: string;
  note_type: OpdNoteType;
  content: string;
  icd_code: string | null;
  icd_description: string | null;
  is_final: boolean;
  created_by_user_id: string | null;
  created_at: string;
}

export interface OpdNoteCreatePayload {
  note_type: OpdNoteType;
  content: string;
  icd_code?: string | null;
  icd_description?: string | null;
}

export interface OpdPrescriptionItem {
  id: string;
  medicine_id: string | null;
  medicine_name: string;
  dosage: string;
  frequency: string;
  duration: string;
  route: MedicineRoute | null;
  instructions: string | null;
  quantity: number | null;
}

export interface OpdPrescription {
  id: string;
  opd_visit_id: string;
  patient_id: string;
  doctor_id: string;
  prescription_number: string;
  prescribed_at: string;
  status: OpdPrescriptionStatus;
  notes: string | null;
  items: OpdPrescriptionItem[];
}

export interface OpdPrescriptionItemCreatePayload {
  medicine_id?: string | null;
  medicine_name: string;
  dosage: string;
  frequency: string;
  duration: string;
  route?: MedicineRoute | null;
  instructions?: string | null;
  quantity?: number | null;
}

export interface OpdPrescriptionCreatePayload {
  notes?: string | null;
  items: OpdPrescriptionItemCreatePayload[];
}
